source/credentials_provider_ecs.c

```diff
@@ -430,9 +432,58 @@ static void s_ecs_query_task_role_credentials(struct aws_credentials_provider_ec
         s_ecs_finalize_get_credentials_query(ecs_user_data);
     }
 }
+/*
+ *  The host must use either HTTPS or the resolved IP address must satisfy one of the following:
+ *   1. within the loopback CIDR (IPv4 127.0.0.0/8, IPv6 ::1/128)
+ *   2. corresponds to the ECS container host 169.254.170.2
+ *   3. corresponds to the EKS container host IPs (IPv4 169.254.170.23, IPv6 fd00:ec2::23)
+ */
+static bool s_is_valid_remote_host_ip(
+    struct aws_credentials_provider_ecs_user_data *ecs_user_data,
+    struct aws_http_connection *connection) {
+    struct aws_credentials_provider_ecs_impl *impl = ecs_user_data->ecs_provider->impl;
+
+    if (impl->is_https) {
+        return true;
+    }
+
+    bool result = false;
+
+    const struct aws_byte_cursor address =
+        aws_byte_cursor_from_c_str(aws_http_connection_get_remote_endpoint(connection)->address);
```

Comment:
This technique won't work with many other HTTP clients. It's uncommon that you have a chance to inspect the IP before making the actual request.

just asking the host_resolver directly would be more portable

I get that you're trying to avoid additional calls to the resolver. You could do like the Rust SDK and [only validate](https://github.com/awslabs/aws-sdk-rust/blob/29883b151225f7b27a22e4d343585d2dc8207521/sdk/aws-config/src/ecs.rs#L176-L178) when full-uri is passed in by the user.

Is the fact that this relies on the HTTP client what's forcing you to add a 2nd way of mocking HTTP calls?